character_map.h

```diff
@@ -0,0 +1,91 @@
+/*
+* Copyright:  Copyright Dimakopoulos Theodoros 2019
+* License:    $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+* Authors:    Dimakopoulos Theodoros
+*/
+#pragma once
+/*
+The character-properties map array.
+
+A character may have multiple properties,
+such as being allowed in a DIL program or
+representing a decimal digit (0, 1, 2, ...9).
+
+ch_map[c] & CM_DIL_CHAR
+will be true if c is a DIL character.
+ch_map[c] == CM_DIL_CHAR
+will be true if the ONLY property is that.
+ch_map[c] == CM_DIL_CHAR + CM_DECIMAL
+will be true if c is DIL and a decimal digit (of course if it is
+decimal it will also be a DIL character. Prefer ch_map[c] & CM_DECIMAL)
+*/
+// CM for 'C'haracter 'M'ap. This way we avoid namespaces.
+const int CM_NON_DIL = 0;
+const int CM_DIL_CHAR = 1;                 // character allowed in DIL
+const int CM_BLANK = 2;                    // space, new-line or tab
+const int CM_IDENTIFIER = 4;               // if could be in an ident.
+const int CM_OCTAL = 8;                    // '0' to '7'
+const int CM_DECIMAL_NOT_OCTAL = 16;       // '8' or '9'
+const int CM_HEXADECIMAL_NOT_DECIMAL = 32; // 'a' to 'f' and 'A' to 'F'
+const int CM_DECIMAL = 16 + 8;             // two properties
+const int CM_HEXADECIMAL = 32 + 16 + 8;    // three properties

```

Comment:
Ok, first of all, you don't need to provide properties that given other properties, can be tested. For example, `CM_DECIMAL_NOT_OCTAL` can be tested as `CM_DECIMAL && ~CM_OCTAL`.
Other than that, the usual trick is sort of what you went for with the bitwise operations, taking advantage of powers of 2. To make it more clear, you can do something like:
```cpp
enum {
CM_NON_DIL = 1,
CM_DIL_CHAR = CM_NON_DIL << 1,
CM_BLANK = CM_DIL_CHAR << 1,
...
```
And generally, as always the advice is make the code as simple as possible. We don't need these complicated cases. All these can be tested with simple combinations between simple cases. For example, `CM_DECIMAL` made the code more unpredictable. Below, by looking at the code, it seems that e.g. `6` is octal but not decimal, which is not true.
One has to pretty much do a very good guess that this is handled with this weird addition.

Reply:
Woops! Somehow I didn't think of that.